Compiled bindings

jkoritzinsky opened the pull request:
## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->
This PR adds an implementation of compile-time verified bindings. The binding paths are verified at compile time and do not use reflection at runtime.

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->
Currently, all bindings are resolved at runtime via reflection.

## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->
By using the `{CompiledBinding}` markup extension, users can have their bindings be pre-compiled, improving performance and helping ensure correctness.

## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->
This is implemented by using the compile-time `IPropertyInfo` interface in #2659 and custom XamlIl transformation passes.

## Checklist

- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Breaking changes
<!--- List any breaking changes here. When the PR is merged please add an entry to https://github.com/AvaloniaUI/Avalonia/wiki/Breaking-Changes -->

Moves most members of `Avalonia.Markup.Data.Binding` to new `Avalonia.Markup.Data.BindingBase` class.

## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->

Blocked by kekekeks/XamlIl#2
Includes #2659 
Contributes to #2359 

This is still a WIP but I'm putting it up for early review if people want to take a look early.

kekekeks committed:
Added IPropertyInfo and use it in IProvideValueTarget when possible

kekekeks committed:
Implemented PropertyPathGrammar

kekekeks committed:
Compiled property paths

kekekeks committed:
Missing file

kekekeks committed:
Do not crash

jkoritzinsky committed:
Convert current binding parsing to create an AST.

jkoritzinsky committed:
Add new AST node generated from x:DataContextType. To be used as part of compiled bindings.

jkoritzinsky committed:
Finish initial implementation of compiled bindings. Still some bugs to work out, but we have something that works now.

jkoritzinsky committed:
Fix stream bindings.

jkoritzinsky committed:
Traverse transform nodes in the correct order when setting a binding source value.

jkoritzinsky committed:
Fix handling of Not nodes.

jkoritzinsky committed:
Make SourceMode internal since it doesn't need to be public.

jkoritzinsky committed:
Add additional test for IObservable compiled bindings.

jkoritzinsky committed:
Change the observable stream binding in BindingDemo to be a compiled binding.

jkoritzinsky committed:
Remove unneeded alias.

jkoritzinsky committed:
Fix namescope searching to recursively search parent namescopes up to the root namescope and never find a name in a namescope that is not a direct ancestor of the current namescope.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/AvaloniaUI/Avalonia into compiled-bindings

jkoritzinsky commented on src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/CompiledBindings/CompiledBindingPath.cs:
```diff
@@ -0,0 +1,179 @@
+
+namespace Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings
+{
+    public class CompiledBindingPath
```
I think its better for them to be separate since in Avalonia, our binding syntax is much more expressive than a simple property path.

stevenbrix replied:
How is the binding syntax more expressive? It would be cool if any improvements Avalonia makes could make their way back to UWP/WPF.

grokys replied:
@stevenbrix mostly we have some syntactic sugar around binding to controls:

http://avaloniaui.net/docs/binding/binding-to-controls

And you can also bind to the result of a `Task` or `IObservable`:

http://avaloniaui.net/docs/binding/binding-to-tasks-and-observables

grokys replied:
Oh and you can also negate values:

http://avaloniaui.net/docs/binding/converting-binding-values#negating-values

jkoritzinsky committed:
Update XamlIl

jkoritzinsky committed:
Handle null namescope in binding

jkoritzinsky commented on src/Markup/Avalonia.Markup.Xaml/XamlIl/CompilerExtensions/Transformers/AvaloniaXamlIlDataContextTypeTransformer.cs:
```diff
@@ -0,0 +1,86 @@
+                        {
+                            calculatedDataContextTypeNode = new AvaloniaXamlIlDataContextTypeMetadataNode(on, pv.Values[0].Type.GetClrType());
+                        }
+                        else if(pv.Values[0].Type.GetClrType().Equals(context.GetAvaloniaTypes().CompiledBindingExtension)
```
Even if we pre-parse the compiled bindings before evaluating, that won’t allow us to infer DataContext’s of referenced elements without a multi-pass transformer. Users could set it up that control A’s data context depends on B’s data context which depends on C’s data context in such a way that our current DFS transformer algorithm (or a BFS) would be unable to resolve in a single pass.

I’d rather leave that to a second PR since this one is already very long.

kekekeks replied:
You can save arbitrary data in XamlIlTransformationContext via SetItem/GetItem. So one the first transformer would populate some data structure with the full list of bindings, then another one can manipulate the data as it sees fit. Another type of transformers that accepts the whole document instead of a tree node might better suit that task.


It's fine to leave that to a separate PR.

jkoritzinsky commented on src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/CompiledBindings/NotifyingPropertyInfoHelpers.cs:
```diff
@@ -0,0 +1,255 @@
+            _base.Set(target, value);
+        }
+
+        public void OnPropertyChanged(object target, EventHandler handler)
```
That's a TODO that I'm working on.

jkoritzinsky commented on src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/CompiledBindings/NotifyingPropertyInfoHelpers.cs:
```diff
@@ -0,0 +1,255 @@
+
+namespace Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings
+{
+    public static class NotifyingPropertyInfoHelpers
```
I'm going to work on a slightly different design for this portion since I had a few ideas going when I first wrote it and ended up with a weird amalgamation of them.

jkoritzinsky committed:
Change property notification management to directly create property accessors. Fix indexer bindings for arrays.

PR Feedback.

jkoritzinsky committed:
Fix BindingDemo

jkoritzinsky committed:
Fix chaining in ExpressionParser.

jkoritzinsky committed:
Fix chain building.

jkoritzinsky committed:
Fix failing tests

jkoritzinsky committed:
Provide helper type instead of manually emitting closure type into assembly.

jkoritzinsky committed:
Merge branch 'master' into compiled-bindings

jkoritzinsky committed:
Revert "Provide helper type instead of manually emitting closure type into assembly."

This reverts commit 2832462a8565891da763484f2b953344f3a3f796.

jkoritzinsky commented:
There's a test failure for array bindings on Mono on macOS that I still need to take a look at, but other than that, this is ready for review.

danwalmsley commented:
@jkoritzinsky do we expect this to replace {Binding } eventually, or will people need to use {CompiledBinding } always?

jkoritzinsky commented:
I think we should keep Binding and  `{CompiledBinding}` separate since CompiledBinding is not fully backward-compatible with regular bindings. Same way that x:Bind didnt replace Binding in UWP.

danwalmsley commented:
> I think we should keep Binding and `{CompiledBinding}` separate since CompiledBinding is not fully backward-compatible with regular bindings. Same way that x:Bind didnt replace Binding in UWP.

Fair enough, I can see that since you have to declare x:DataContextType, so it behaves a bit different to normal bindings!

jkoritzinsky commented:
That's the main difference. Additionally, compiled bindings only support the shorthand syntax for RelativeSource bindings, which simplifies the implementation drastically.

kekekeks commented:
> only support the shorthand syntax for RelativeSource bindings, which simplifies the implementation drastically.

That can be fixed by adding a transformation step that converts RelativeSource to the shorthand one.

I'd like to replace the regular Binding completely at some point, especially since people coming from other XAML frameworks will try to use this name first.

kekekeks commented:
We should probably add `CompileBindings` MSBuild property and `x:CompileBindings` directive to replace regular bindings with compiled ones and include that in 0.9 release. For 0.10 we can make both the msbuild property be true by default.

danwalmsley commented:
@jkoritzinsky @kekekeks could we make Binding be compiled if the following conditions are met:

1) the x:DataContextType is set.
2) use the compiled binding when DataContext type matches (this check can be done once)

Then if any of those conditions are not met, fallback to reflective binding.

and use {CompiledBinding } to force compiled?

So you kind of get a best of both worlds... optimized if it can

kekekeks commented:
Mkay, here the proposal:

1) `{CompiledBinding ...}` is always compiled
2) `{ReflectionBinding ...}` always uses the current implementation
3) `{Binding ...}` is mapped to one of those based on `CompileAvaloniaBindings` MSBuild property and `x:CompileBindings` directive which overrides the property and enforces the corresponding policy for a particular file
4) In 0.9 `CompileAvaloniaBindings` will be `false` by default, so people could opt-in
5) In 0.10 `CompileAvaloniaBindings` will be `true` by default, so people could opt-out

jkoritzinsky committed:
Fix failing test on Mono runtimes.

jkoritzinsky commented:
Before we map `Binding` to something other than the `Avalonia.Markup.XamlIl.MarkupExtensions.BindingExtension` type or the `Avalonia.Markup.Data.Binding` type, we should make sure that we have an alternative for C# code that's also compiled (preferably building off the expression-tree parsing support in ExpressionObserver).

jkoritzinsky committed:
Merge branch 'master' of https://github.com/AvaloniaUI/Avalonia into compiled-bindings

Commit:
Merge branch 'master' into compiled-bindings

jkoritzinsky committed:
Implement smart inference for DataTemplates. Fix calculated DataContext types for DataContext bindings.

jkoritzinsky committed:
Merge branch 'compiled-bindings' of https://github.com/jkoritzinsky/Avalonia into compiled-bindings

danwalmsley commented:
@jkoritzinsky xamarin published some blog posts
https://devblogs.microsoft.com/xamarin/compiled-bindings-xamarin-forms/?utm_source=vs_developer_news&utm_medium=referral

https://docs.microsoft.com/en-gb/xamarin/xamarin-forms/app-fundamentals/data-binding/compiled-bindings?WT.mc_id=compiledbindings-xamarinblog-jamont#performance

I wonder if we can match them

I particularly like use of `x:DataType` and  `[assembly:XamlCompilation(XamlCompilationOptions.Compile)]`

what do you guys think?

jkoritzinsky commented:
Changing to `x:DataType` is a 1-line fix that we can do easily. The `[assembly:XamlCompilation(XamlCompilationOptions.Compile)]` attribute is used in Xamarin.Forms to enable compiled XAML as a whole, not just compiled bindings.

jkoritzinsky committed:
Fix test failures. Ensure that BindingExtension is detected as such an not mistaken for the actual DataContext type.

jkoritzinsky committed:
x:DataContextType -> x:DataType to match Xamarin.Forms.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/AvaloniaUI/Avalonia into compiled-bindings

jkoritzinsky committed:
Fix failing test.

jkoritzinsky committed:
Throw exception if trying to compile a binding under an element with its DataContext bound via a non-compiled binding.

grokys committed:
Merge branch 'master' into compiled-bindings

grokys commented:
Reviewing now with by trying to integrate my "bindings-as-expressions" branch , going to update this comment with my findings as I go along:

- `Text="{CompiledBinding Path=StringValue}"` doesn't work: `Avalonia error XAMLIL: Unable to find suitable setter or adder for property Path of type Avalonia.Markup.Xaml:Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindingExtension for argument System.Runtime:System.String`

jkoritzinsky committed:
Fix assigning binding paths via the property.

jkoritzinsky committed:
Merge branch 'compiled-bindings' of https://github.com/jkoritzinsky/Avalonia into compiled-bindings

jkoritzinsky commented:
@grokys fixed the first item on your list.

It copies itself in ProvideValue because the provided value is CompiledBinding+default anchor. I copied the design from Binding vs BindingExtension. I'm also willing to change it.

stevenbrix commented:
![image](https://user-images.githubusercontent.com/9649518/66586843-2f905a80-eb3e-11e9-980d-e4f2d2e724c0.png)

This is exciting! I would love to see this match what Xamarin.Forms has done for compiled bindings. The way that they did compiled bindings is (in my opinion) a better model than what we did for UWP. Hindsight is always 20-20 I guess :).

@kekekeks I heard on the Hansleminutes podcast with @grokys that you own the XAML compiler for this project, and I'd love to work with you guys closer if that interests you. We're working on refactoring the UWP markup compiler in a way that we'll share it across WPF, Xamarin.Forms, and WinUI. We want to design our tooling in a way that we can close the gap on the differences between the different XAML dialects and where future improvements will reach all our customers. 

> I think we should keep Binding and {CompiledBinding} separate since CompiledBinding is not fully backward-compatible with regular bindings. Same way that x:Bind didnt replace Binding in UWP.

@jkoritzinsky this is something that we are working on with `x:Bind` and it's unfortunate that it doesn't completely replace `{Binding}`. I personally like that Xamarin.Forms didn't make them separate.

kekekeks commented:
@stevenbrix [XamlIl](https://github.com/kekekeks/XamlIl) was designed to be platform-agnostic, so it doesn't rely on anything specific to a particular UI framework. That's one of the reasons it lives in a separate repository. So every interface/type/attribute is [configurable](https://github.com/kekekeks/XamlIl/blob/master/src/XamlIl/Transform/XamlIlLanguageTypeMappings.cs).
Things that are specific to Avalonia are implemented via [a separate set of AST transformations](https://github.com/AvaloniaUI/Avalonia/tree/master/src/Markup/Avalonia.Markup.Xaml/XamlIl/CompilerExtensions/Transformers) that live in Avalonia repo.

So it should be easy to add other XAML dialects.

On the other hand, the intention was to make XAML loading to be completely reflection-free, so generated code would be 100% linkable. So there are several things that might prevent it from be used in WPF:

- IXamlSchemaContextProvider: we don't have a schema context at run time
- IAmbientProvider - we don't have "xaml type system" at run time, only plain CLR types


See the README [there](https://github.com/kekekeks/XamlIl), it's mostly up to date.

stevenbrix commented:
> On the other hand, the intention was to make XAML loading to be completely reflection-free, so generated code would be 100% linkable. So there are several things that might prevent it from be used in WPF

Could there be a WPF specific plug-in that enables those scenarios, or are they broken by design?

The overall architecture and idea of XamlIL looks really promising, what made you want to produce IL instead of code that gets generated into IL by the C# compiler?

jkoritzinsky commented:
I'll start taking a stab at supporting precompilation of more of the features of the classic `{Binding}` syntax so we can unify on having XamlIL transform `{Binding}` to `{CompiledBinding}`.

> > On the other hand, the intention was to make XAML loading to be completely reflection-free, so generated code would be 100% linkable. So there are several things that might prevent it from be used in WPF
> 
> Could there be a WPF specific plug-in that enables those scenarios, or are they broken by design?

From my understanding, they could be implemented, but you'd basically have to emit all of the data into the IL so that the `IServiceProvider` implementation would have enough info to correctly implement the interfaces. With enough work with custom AST passes and a custom `IServiceProvider` implementation, it should be possible to implement these interfaces. We just don't need them for Avalonia, so the compiler doesn't support them currently.

> 
> The overall architecture and idea of XamlIL looks really promising, what made you want to produce IL instead of code that gets generated into IL by the C# compiler?

We produce IL for a few reasons:

- Compatibility with all .NET languages. We have a set of users that use F#, and we don't want to lock them out from being able to use Avalonia's XAML engine.
- Unspeakable names: By emitting IL, we can emit type names that are illegal in C# (and likely F#) but are valid in IL. As a result, we can ensure that we are extremely unlikely to have a type name conflict (we could also change what we emit if we detected a conflict, but we currently don't do that).
- Our runtime and compilation-time XAML engines are exactly the same except for the IL emit backend. This allows people to (if they want to) parse XAML at runtime the same way as if it were at compile time. It also makes testing the custom XamlIL transform passes in Avalonia significantly easier since we can use the runtime emit engine and write regular Xunit unit tests for most tests.

@stevenbrix I also work at Microsoft (specifically on interop in .NET), so if you want to chat more offline, you can put some time on my calendar and we can talk about Avalonia and XamlIL things.

kekekeks commented:
@stevenbrix 

>Could there be a WPF specific plug-in that enables those scenarios

` IAmbientProvider` would be essentially an expensive emulation layer that *might* need quite a lot of extra info saved during compilation due to the usage of [XamlMember](https://github.com/AvaloniaUI/Avalonia/blob/9910610573e11a7b91ea9094faad88462f9a2bb9/src/Markup/Avalonia.Markup.Xaml/XamlIl/Runtime/IAvaloniaXamlIlParentStackProvider.cs) APIs.

If possible, WPF should make some of the markup extensions more compiler-friendly. In the process of porting Avalonia XAML loader to use XamlIl, I've had several helper methods that [were using](https://github.com/AvaloniaUI/Avalonia/commit/76f9c8fdc5831c4571147cf098b082741ea49f44#diff-63ca212732140016e660524bde7044ddL25) either [IAmbientProvider](https://docs.microsoft.com/en-us/dotnet/api/system.xaml.iambientprovider?view=netframework-4.7.2) from Portable.Xaml or way cheaper [IParentStackProvider](https://github.com/AvaloniaUI/Avalonia/blob/9910610573e11a7b91ea9094faad88462f9a2bb9/src/Markup/Avalonia.Markup.Xaml/XamlIl/Runtime/IAvaloniaXamlIlParentStackProvider.cs) from XamlIl depending on which was available.

Surprisingly no code bases were broken by removing IAmbientProvider, but that most certainly won't be the case for WPF.

jkoritzinsky committed:
Fix ElementName bindings.

jkoritzinsky committed:
Enable transforming old-style RelativeSource/ElementName properties in CompiledBindings to element path nodes during XamlIl compilation.

jkoritzinsky committed:
Support synthetic Source property.

jkoritzinsky committed:
Merge branch 'master' into compiled-bindings

jkoritzinsky commented:
I've updated this PR to support the full syntax of the traditional `{Binding}` markup extension in the `{CompiledBinding}` markup extension.  

@kekekeks this change makes it possible to just swap out compiled bindings for traditional bindings during XamlIL compilation. What's the design you were thinking for automatically converting reflection-based bindings to compiled bindings?

kekekeks commented:
>What's the design you were thinking for automatically converting reflection-based bindings to compiled bindings

1) There should be `ReflectionBinding` and `CompiledBinding` that always correspond to compiled or reflection based bindings. Basically that's a way for the user to force the binding to be the desired type
2) We treat regular `{Binding}` as either reflection or compiled binding based on the preference for the current file
3) The preference could be set via `AvaloniaCompileBindings` MSBuild property project-wise and `x:CompileBindings` on the root element of the file

The way that could be implemented in the compiler is some special handling for `CompiledBinding` with AST rewrite step that 

- converts `Binding` to `CompiledBinding` or leaves it as is
- converts `ReflectionBinding` to `Binding`

stevenbrix commented:
> What's the design you were thinking for automatically converting reflection-based bindings to compiled bindings

@kekekeks, @jkoritzinsky can you guys follow the same principle that Xamarin.Forms does, which is via the presence of `x:DataType`. Any `{Binding}` in that subtree is converted to Compiled. Unless there is serious precedent to break that pattern, I think it is a reasonable and simple way of determining when to do it

jkoritzinsky commented:
@stevenbrix I prefer @kekekeks design for a few reasons. The main reason is this implementation of compiled bindings is significantly smarter than the Xamarin.Forms one in that it automatically does common inference to make sure the expected DataType is up to date.

Based on documentation, the Xamarin.Forms implementation doesn't automatically infer the `x:DataType` value when the `BindingContext` property is set. Nor does it automatically infer the correct type for a typeless `DataTemplate` within an `ItemsControl` whose `ItemsSource` property is set to a collection. If you re-bind the `BindingContext`, you'll get incorrectly compiled bindings. Depending on your property names, it might successfully compile and then fail (likely cryptically) at runtime.

This implementation automatically supports that inference. Additionally, this implementation will throw an exception at compile time when it detects that you're using a compiled binding in a context that it cannot correctly run its inference on, avoiding obscure runtime failures. Ours will also support setting the `Source` property on the binding. I also plan to support compiling long-form element-style bindings once I add the opt-in renaming with whatever design we choose.

Additionally, the Xamarin.Forms implementation also has the assembly-wide off switch by turning off (or at least not turning on) compiled XAML. In Avalonia, our XAML is always compiled, so this switch wouldn't make sense in our scenario.

Also, I'm not a big fan of how in the Xamarin.Forms implementation, you have to manually go search back up the logical tree to see if a binding is runtime or compile time. Having the option to be explicit for individual bindings (`ReflectionBinding` and `CompiledBinding`) is really nice.

As a result of having all of this "magic" that the Xamarin.Forms implementation doesn't have, I think it's worth having a slightly different opt-in design.

stevenbrix commented:
> The main reason is this implementation of compiled bindings is significantly smarter than the Xamarin.Forms one in that it automatically does common inference to make sure the expected DataType is up to date.

With UWP we explicitly decided to not use `DataContext` for `x:Bind`. I don't remember the real reason, but my hunch is it was this sort of thing. `DataContext` can change at runtime, which means your compiled binding always has a chance to fail at runtime, which defeats the purpose.

> If you re-bind the BindingContext, you'll get incorrectly compiled bindings. Depending on your property names, it might successfully compile and then fail (likely cryptically) at runtim

Is this what happens with Xamarin.Forms? I'll admit I haven't played around enough to know exactly what they've done. This sort of defeats the purpose of compiled bindings :)

> The main reason is this implementation of compiled bindings is significantly smarter than the Xamarin.Forms one in that it automatically does common inference to make sure the expected DataType is up to date.

> As a result of having all of this "magic" that the Xamarin.Forms implementation doesn't have, I think it's worth having a slightly different, more explicit, opt-in design.

Fair, I'm against magic for sure, but I feel like `x:DataType` is an explicit enough usage. In UWP we use `x:DataType` for compiled bindings in DataTemplates. Validating that the `DataContext` is of the same type is definite goodness, I'm not saying don't do that just because Xamarin.Forms might not.

Is there a particular reason you guys don't want to follow the same model and developer scenario as we did with UWP?

jkoritzinsky commented:
> > The main reason is this implementation of compiled bindings is significantly smarter than the Xamarin.Forms one in that it automatically does common inference to make sure the expected DataType is up to date.
> 
> With UWP we explicitly decided to not use `DataContext` for `x:Bind`. I don't remember the real reason, but my hunch is it was this sort of thing. `DataContext` can change at runtime, which means your compiled binding always has a chance to fail at runtime, which defeats the purpose.

Generally you don't change your root `DataContext` type so I'm not sure if that's it. UWP's `x:Bind` also has a different default binding mode, which is also odd.

> 
> > If you re-bind the BindingContext, you'll get incorrectly compiled bindings. Depending on your property names, it might successfully compile and then fail (likely cryptically) at runtim
> 
> Is this what happens with Xamarin.Forms? I'll admit I haven't played around enough to know exactly what they've done. This sort of defeats the purpose of compiled bindings :)

From my understanding of their documentation (I haven't run it, just going off their docs), the following code would silently compile incorrectly. The `{Binding Id}` binding will be compiled for `ViewModel.Id` instead of `SubViewModel.Id`.

```csharp
namespace DataBindingDemos
{
class ViewModel
{
     public SubViewModel SubModel { get; set; }
     public string Id { get; set; }
}

class SubViewModel
{
    public int Id { get; set; }
}
}
```
```xml
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms" 
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:local="clr-namespace:DataBindingDemos"
             x:Class="DataBindingDemos.DemoPage"
             x:DataType="local:ViewModel">
   <Button BindingContext="{Binding SubModel}" Text="{Binding Id}" />
</ContentPage>
```

> > The main reason is this implementation of compiled bindings is significantly smarter than the Xamarin.Forms one in that it automatically does common inference to make sure the expected DataType is up to date.
> 
> > As a result of having all of this "magic" that the Xamarin.Forms implementation doesn't have, I think it's worth having a slightly different, more explicit, opt-in design.
> 
> Fair, I'm against magic for sure, but I feel like `x:DataType` is an explicit enough usage. In UWP we use `x:DataType` for compiled bindings in DataTemplates. Validating that the `DataContext` is of the same type is definite goodness, I'm not saying don't do that just because Xamarin.Forms might not.
> 
> Is there a particular reason you guys don't want to follow the same model and developer scenario as we did with UWP?

The differing defaults in UWP with the source and the defaults were basically a no-go. Our design goal has always been more similar to Xamarin.Forms' implementation.

I think a design in the middle where having an explicit `x:DataType` opts-in for the tree below (excluding through deferred content), an `x:DataType="{x:Null}"`  attribute assignment opts-out and `ReflectionBinding` and `CompiledBinding` allow deciding for specific bindings would work for me.

jkoritzinsky commented:
Actually the more I think about it, we can't use `x:DataType` as our diambiguator since we support non-data-typed DataTemplates with an implied type (for ItemsControls with collections bound).

kekekeks commented:
Also, explicit switch to disable reflection-based bindings is a requirement for scenarios like IL linking, aggressive post-build obfuscation and running in CoreRT's reflection-free mode. For these scenarios we need to disable reflection usage with compile-time errors.

stevenbrix commented:
> Actually the more I think about it, we can't use x:DataType as our diambiguator since we support non-data-typed DataTemplates with an implied type (for ItemsControls with collections bound).

@jkoritzinsky can you explain what this means in more details? What does this scenario look like?

> Also, an explicit switch to disable reflection-based bindings is a requirement for scenarios like IL linking, aggressive post-build obfuscation and running in CoreRT's reflection-free mode. For these scenarios we need to disable reflection usage with compile-time errors.

@kekekeks makes sense, I like this idea 👍

jkoritzinsky committed:
Rename BindingExtension->ReflectionBindingExtension.

Implement x:CompileBindings directive to toggle between transforming Binding->ReflectionBindingExtension and Binding->CompiledBindingExtension.

Expose a property on AvaloniaXamlIlCompiler to set the default transformation.

jkoritzinsky committed:
Merge branch 'compiled-bindings' of https://github.com/jkoritzinsky/Avalonia into compiled-bindings

hez2010 commented on src/Avalonia.Base/Data/Core/ClrPropertyInfo.cs:
```diff
@@ -0,0 +1,73 @@
+                return null;
+            var target = Expression.Parameter(typeof(object), "target");
+            var value = Expression.Parameter(typeof(object), "value");
+            return Expression.Lambda<Action<object, object>>(
```
Runtime code-generation is not allowed in AOT compilation. Use of Expression.Lambda().Compile() will cause build failure during AOT.

jkoritzinsky replied:
Expression.Lambda().Compile() has an interpreter fallback built into the BCL. It's not super fast, but it's still AOT-friendly.

hez2010 commented on src/Avalonia.Base/Data/Core/ClrPropertyInfo.cs:
```diff
@@ -0,0 +1,73 @@
+            var target = Expression.Parameter(typeof(object), "target");
+            return Expression.Lambda<Func<object, object>>(
+                    Expression.Convert(Expression.Call(Expression.Convert(target, info.DeclaringType), info.GetMethod),
+                        typeof(object)))
```
Test ok with AOT, but the parameter "target" was missed here.
```suggestion
                        typeof(object)), target)
```

jkoritzinsky committed:
Merge branch 'master' into compiled-bindings

jkoritzinsky committed:
Merge branch 'master' into compiled-bindings

jkoritzinsky committed:
Fix build failure from the merge to master.

jkoritzinsky committed:
Rename PeekOneOrThrow back to Peek.

jkoritzinsky committed:
Merge branch 'master' of https://github.com/AvaloniaUI/Avalonia into compiled-bindings

Commit:
remove un-needed $parent bindings causing bugs in progress bar.

danwalmsley reviewed (approved):
Good work. we have given this some testing and decided to get this into the wild for 0.10 release.

Commit:
Merge branch 'master' into compiled-bindings
